### ISSUE TITLE
Format and value of period dates

### ISSUE THREAD

**Issue by mark-prins** [openmsupply/mobile#3034]
## Describe the bug

The start and end date of periods, shown when creating a new requisition, are both incorrect and in US date format.

### To reproduce

Steps to reproduce the behaviour:

1. Change regional settings on the tablet, so that the timezone is non-NZ ( I selected the region of Côte d'Ivoire => GMT )
2. Create a new program-based supplier requisition
3. When you reach the `period` selection, observe the dates and compare to mSupply

### Expected behaviour

- Date format should be consistent. Currently, mSupply shows DD/MM/YYYY and mobile shows MM/DD/YYYY
- Dates are affected by the timezone difference between server and mobile

### Additional context

![Screen Shot 2020-07-20 at 5 29 13 PM](https://user-images.githubusercontent.com/9192912/87902948-01697f00-caaf-11ea-8210-f76181f4f2f8.png)

**Comment by mark-prins** [comment]
Similar to #2946 ?

**Comment by joshxg** [comment]
For the formatting:

It's formatted from: 

- https://github.com/openmsupply/mobile/blob/f97a07481663f759082e8c70e8fb62e35934c0be/src/database/DataTypes/Period.js#L51

Uses:

- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString

- apparently not a good idea: https://stackoverflow.com/questions/51399551/why-isnt-tolocaledatestring-working-in-react-native-android/51400889

This is the only place it's used. Should use `moment`. Why it's used is because ... 🤷 


For the timezone offset: Seems #2946 added timezone offsets, which we were ignoring previously. I am unsure what I think is the right behaviour but I am leaning toward "if the server sends it with an offset, use it. If we don't want to ever show it, don't send it?" I dunno, I hate dates.

But solution for formatting: Use `moment` to format `Period.toString()` in the usual format `DD/MM/YYYY`. For offset, whatever @mark-prins says?? (:buck:??)

**Comment by mark-prins** [comment]
yeah, thanks Josh. I did think to change to `moment`
and agree. Dates suck.
Ideal solution: convert to a string before sending and pretend that these aren't dates.
I'll have a brief look at less-than-ideal-but-don't-have-to-change-4D options for date munging.

**Comment by mark-prins** [comment]
Oh - further investigation. mSupply is sending the date as UTC, which in this case is very unhelpful.

- set period start date to 7/04/2019 ( server is in NZ timezone.. +13 )
- server sends the string `2019-04-06T11:00:00Z` to mobile
- mobile ( GMT timezone.. +0 ) reads this as 11:00 on the 6/4/19 and shows that date to the user

however, mobile doesn't know which timezone the server is in, so cannot convert the date back to the correct value.
Same issue with all the other dates, surely? I've just checked supplier requisitions, and yes. The date shown on the server is not the date shown on mobile.

**Comment by joshxg** [comment]
Sorry: are we saying we need to make the changes to the server to send an offset?

**Comment by wlthomson** [comment]
Just to clarify, the server is sending the date using the UTC offset, which (as I understand it) is a "neutral" format (server converts local time to UTC, mobile converts UTC to local time). So I don't think the problem here is UTC. 

I've done a few things in this area, so I'll look into this now. My initial thoughts are inline with what @joshxg has said about using `moment` instead of date methods. I'm fairly sure this will be all that's needed (shouldn't be any server-side changes unless mSupply is parsing incoming dates weirdly).

**Comment by mark-prins** [comment]
that's one option - usually there's less trouble just sending as string rather than dates e.g. sending `07/04/2019`

> I'm fairly sure this will be all that's needed

Not convinced - the server sends ` 2019-04-06T11:00:00Z` and mobile should display `07/04/2019` but how would mobile know to add 13h to the UTC date?
The server would have to send the +13 or send the actual date without subtracting an unspecified number of hours first.

**Comment by joshxg** [comment]
more confirming needed for me sorry... ! :( I have a feeling we might have different answers for the following, or I am just crazy?

what exactly is the agreed and desired behavior in the following example:

- mSupply server is 12 hours ahead of a tablet running mSupply mobile.
- mSupply server has a requisition created @ 11am on the first of February for the mobile site.
- Mobile site receives the record

The Q: What is the date created showing on the requisition on mobile: 31/1/2020 (@ 1AM), or 1/2/2020 (@ 11AM)?

**Comment by wlthomson** [comment]
@mark-prins As I understand it, desktop knows it's own timezone, so when the date is converted to UTC, the date/time is adjusted. Mobile also knows it's own timezone, so it will convert the UTC so the equivalent time in its timezone (i.e. UTC is a "common" time format that is used to express a date without timezone offsets). 

@joshxg My understanding is that each device should show the date in it's own timezone. To conceptualise it, a requisition is  created at a certain time in UTC (no matter the timezone), mobile and desktop apply their own timezones, but the UTC date/time is consistent (i.e. if mobile and desktop both converted the requisition date/time to UTC it would be the same).

Does that make sense?

**Comment by mark-prins** [comment]
@joshxg currently - the date shown on mobile is 31/101/2020 
If we do show time somewhere then it would be 23:00 / 11pm

@wlthomson so, yeah. As it stands, the date is technically correct, in that it specifies a point in time which is consistent.
The problem though is the picture above in 'Additional Details'.

The client created a period that starts on date1 and ends on date2. When viewed on mobile, the date shown is date1 - 1 day and date2 - 1 day. They are not happy about that. Trying to argue that date1 is in a different timezone and technically refers to the same point in time.. is not going to fly.

Really, someone should just change the timezone on the server! and then re-initialise all of the tablets. Not sure that this is possible, or even safe - though it should be ok.

**Comment by wlthomson** [comment]
@mark-prins So just to clarify, is the mobile timezone different from the server? If that's the case, then I think the timezone needs to be changed wherever it needs to be! 

I don't think it's a good idea to work around this by ignoring timezones; in fact this was a previous bug which we had to fix (so even if made the client happy, we would end up regressing another bug... and likely causing a lot of other problems). 

TLDR; I think this is a config issue, not a bug?

**Comment by joshxg** [comment]
Closed by #3036
